File: week5/python/WordSearch_79.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        
        visited = set()
        def dfs(x: int, y: int, index: int) -> bool:
            if index == len(word):
                return True
            if x < 0 or y < 0 or x >= len(board) or y >= len(board[x]) \
                or (x,y) in visited or board[x][y] != word[index]:
                return False
            
            index +=1
            visited.add((x,y))
            if dfs(x-1,y,index) or dfs(x+1,y,index) \
                or dfs(x,y-1,index) or dfs(x,y+1,index):
                return True
            
            visited.remove((x,y))
            return False
        
        for r in range(len(board)):
            for c in range(len(board[r])):
                if (r,c) not in visited and dfs(r,c,0):
                    return True
                
        return False

File: week5/java/WordSearch_79.java
class Solution {
    public boolean exist(char[][] board, String word) {
        if (word == null || word.length() == 0 || board == null || board.length == 0)
            return true;
        
        boolean[][] visited = new boolean[board.length][board[0].length];
        char[] charArr = word.toCharArray();
        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[i].length; j++) {
                if (isWordExist(board, charArr, i, j, 0, visited))
                    return true;
            }
        }
        return false;
    }
    
    private boolean isWordExist(char[][] board, char[] charArr, int r, int c, int index, boolean[][] visited) {
        if (index == charArr.length)
            return true;
        
        if (r < 0 || c < 0 || r >= board.length || c >= board[r].length || board[r][c] != charArr[index] || visited[r][c])
            return false;
        
        index += 1;
        visited[r][c] = true;
        boolean found =  isWordExist(board, charArr, r + 1, c, index, visited)
                        || isWordExist(board, charArr, r, c + 1, index, visited)
                        || isWordExist(board, charArr, r - 1, c, index, visited)
                        || isWordExist(board, charArr, r, c - 1, index, visited);
        visited[r][c] = false;
        return found;
    }
    
}
